pkg/import/importer.go

```diff
@@ -0,0 +1,467 @@
+		k8sClient:           k8sClient,
+		printer:             printer,
+		waiter:              waiter,
+		imageFetcher:        fetcher,
+		timestampProvider:   timestampProvider,
+		clusterStackFactory: clusterstack.NewFactory(printer, relocator, fetcher),
+		clusterStoreFactory: clusterstore.NewFactory(printer, relocator, fetcher),
+	}
+}
+
+func (i *Importer) ReadDescriptor(descriptorReader io.Reader) (DependencyDescriptor, error) {
+	return readDescriptor(descriptorReader)
+}
+
+func (i *Importer) ImportDescriptor(ctx context.Context, keychain authn.Keychain, kpConfig config.KpConfig, descriptorReader io.Reader) error {
+	descriptor, err := readDescriptor(descriptorReader)
+	if err != nil {
+		return err
+	}
+
+	rDescriptor, err := i.relocateDescriptor(ctx, keychain, kpConfig, i.timestampProvider.GetTimestamp(), descriptor)
+	if err != nil {
+		return err
+	}
+
+	if rDescriptor.lifecycle != nil {
+		if err := i.updateLifecycleConfigMap(ctx, rDescriptor.lifecycle); err != nil {
+			return err
+		}
+	}
+
+	storeToGeneration := map[string]int64{}
+	for _, store := range rDescriptor.clusterStores {
+		gen, err := i.saveClusterStore(ctx, store)
+		if err != nil {
+			return err
+		}
+
+		storeToGeneration[store.Name] = gen
+	}
+
+	stackToGeneration := map[string]int64{}
+	for _, stack := range rDescriptor.clusterStacks {
+		gen, err := i.saveClusterStack(ctx, stack)
+		if err != nil {
+			return err
+		}
+
+		stackToGeneration[stack.Name] = gen
+	}
+
+	for _, builder := range rDescriptor.clusterBuilders {
+		if err := i.saveClusterBuilder(ctx, storeToGeneration, stackToGeneration, builder); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (i *Importer) ImportDescriptorDryRun(ctx context.Context, keychain authn.Keychain, kpConfig config.KpConfig, descriptorReader io.Reader) error {
+	descriptor, err := readDescriptor(descriptorReader)
+	if err != nil {
+		return err
+	}
+
+	_, err = i.relocateDescriptor(ctx, keychain, kpConfig, i.timestampProvider.GetTimestamp(), descriptor)
+	return err
+}
+
+func (i *Importer) Objects() []runtime.Object {
+	return i.objs
+}
+
+func (i *Importer) relocateDescriptor(ctx context.Context, keychain authn.Keychain, kpConfig config.KpConfig, ts string, descriptor DependencyDescriptor) (relocatedDescriptor, error) {
+	var (
+		updatedLifecycle *corev1.ConfigMap
+		err              error
+	)
+
+	if descriptor.HasLifecycleImage() {
+		updatedLifecycle, err = i.relocateLifecycle(ctx, keychain, kpConfig, ts, descriptor.GetLifecycleImage())
+		if err != nil {
+			return relocatedDescriptor{}, err
+		}
+		i.trackObj(updatedLifecycle)
+	}
+
+	clusterstores := make([]*v1alpha1.ClusterStore, 0)
+	for _, clusterStore := range descriptor.ClusterStores {
+		rStore, err := i.constructClusterStore(ctx, keychain, kpConfig, clusterStore)
+		if err != nil {
+			return relocatedDescriptor{}, err
+		}
+		rStore.Annotations = k8s.MergeAnnotations(rStore.Annotations, map[string]string{"kpack.io/import-timestamp": ts})
```

Comment:
Should this take place in construct store?

Reply:
it could but then id have to do it, but then id have to do it on lines 247 and 254 (to handle existing vs new stores) so i figured doing it here was simpler